analysis/dp_strategy_selector.py

```diff
@@ -0,0 +1,197 @@
+# Copyright 2023 OpenMined.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Choosing DP Strategy (i.e. noise_kind, partition selection strategy etc)
+based on contribution bounding params."""
+from dataclasses import dataclass
+from typing import Optional
+
+import pipeline_dp
+from pipeline_dp import aggregate_params
+from pipeline_dp import dp_computations
+from pipeline_dp import input_validators
+
+
+@dataclass
+class DPStrategy:
+    """Represents the chosen DP methods."""
+    noise_kind: Optional[pipeline_dp.NoiseKind]
+    partition_selection_strategy: Optional[
+        pipeline_dp.PartitionSelectionStrategy]
+    post_aggregation_thresholding: bool
+
+
+class DPStrategySelector:
+    """Chooses DPStrategy based on DP budget, computation sensitivites etc
+
+    It chooses noise_kind to minimize the noise std deviation.
+    For non-public partitions it chooses partition selection strategy to
+    minimize threshold. For more details see docstring to
+    select_partition_selection().
+    """
+
+    def __init__(self, epsilon: float, delta: float,
+                 metric: Optional[pipeline_dp.Metric],
+                 is_public_partitions: bool):
+        input_validators.validate_epsilon_delta(epsilon, delta,
+                                                "DPStrategySelector")
+        if delta == 0 and is_public_partitions:
+            raise ValueError(
+                "DPStrategySelector: when public_partitions are used, delta must be positive"
+            )
```

Comment:
Why? Maybe they'll use Laplace & public partitions => delta = 0?

Reply:
thanks, it's vice versa :)